proposals/sized-binding-arrays.md

```diff
@@ -0,0 +1,114 @@
+
+A new `binding_array<T, N>` type is added that can only be used as a binding and not created directly in shaders.
+
+ - `N` is the size of the array and must be a compile-time constant value or (optionally) an override constant.
+   It is validated to less than or equal to the `arraySize` of the respective `GPUBindGroupLayoutEntry`.
+ - `T` is the type of the binding this array is used for, which is validated against the rest of the contents of the `GPUBindGroupLayoutEntry`.
+
+(TODO for WGSL, can `binding_array` be passed around in function parameters?)
+
+New array access expressions are added for `binding_array`:
+
+ - `e: binding_array<T, N> and i: i32 or u32` -> `e[i]: T`. The result is the value of the `i`th element of the array value `e`.
+    If `i` is outside of the range `[0, N-1]`:
+    
+    - It is a shader-creation error if `i` is a const-expression.
+    - It is a pipeline-creation error if `i` is an override-expression.
+    - Otherwise, any element in the array may be returned.
+
+ - Similar rules for getting a reference to an array element from a memory view to a binding array.
+
+Optionally validation rules for the uniformity, or const-expression-ness of indices into the access expressions are added.
+
+## Extra considerations
+
+WebGPU has composite binding objects with `GPUExternalTexture` that takes multiple slots for limits, and may be represented by multiple bindings in implementations.
+The proposal is to allow `binding_array` for `GPUExternalTexture` as well since it can be translated into per-binding types arrays in shaders.
+
+Texel buffers are a potential future addition to WebGPU and should be considered when deciding which binding types can be arrayed.
+
+## Choices for options outlined above.
+
+To be confirmed:
+
+ - Metal and D3D12 allow non-uniform indexing of bindings.
+ - OpenGL ES allows non-uniform indexing of at least textures.
+ - Vulkan has feature for each individual binding type to allow uniform and non-uniform indexing.
+ - Vulkan uniform indexing of everything but texel buffers is ubiquitous (loses only some LLVMPipe reports in the vulkan gpuinfo query)
+ - Vulkan uniform indexing of texel buffers is added in Vulkan 1.2, but very prevalent there (apart from some Intel devices?)
+ - Vulkan non-uniform indexing of all kinds of bindings is very prevalent but not ubiquitous.
+
+Given that a reasonable workaround exists to support uniform and non-uniform indexing on devices that don't support it, and that the workaround has been used with great success in WebGL, the proposal is to allow non-uniform indexing of all kinds of bindings.
+The workaround looks like the following:
+
+```wgsl
+// Replaces
+var e = myBindingArray[i];
+
+// With
+var e = indexMyBindingArray(i);
+
+fn indexMyBindingArray(i) -> T {
+  switch (i) {
+    case 0 { return myBindingArray[0]; }
+    case 1 { return myBindingArray[1]; }
+    // ...
+    case N-1 { return myBindingArray[N-1]; }
+  }
+  return myBindingArray[0];
+}
```

Comment:
I think we need to investigate this more thoroughly for buffers.

In SPIR-V, you'd naturally want to return a pointer to the Block, but that requires VariablePointers. Unfortunately, VariablePointers only work for Workgroup and StorageBuffer storage classes (not Uniform). Even inlining the function doesn't help because you'd need to either store the pointer (requires variable pointers) or use an OpPhi (requires variable pointers). So it's possible there's no way to polyfill this without uniform indexing for uniform buffers.

D3D might have similar problems for buffers. Textures definitely work in both languages.

Reply:
Looking at gpuinfo, I think uniform dynamic indexing is ok at [98%](https://vulkan.gpuinfo.org/listdevicescoverage.php?feature=shaderUniformBufferArrayDynamicIndexing&platform=all) and storage buffer is probably ok too at [~95%](https://vulkan.gpuinfo.org/listdevicescoverage.php?feature=shaderStorageBufferArrayDynamicIndexing&platform=all). So dynamically uniform should be ok.

Reply:
Thinking a bit more in case D3D is a problem. The switch could actually be inline and have the memory access hoisted into it. It might be ugly and bloat the code, but it could be made to work.